### ISSUE TITLE
Optional property could be undefined even after explicitly checking for truthiness

### ISSUE THREAD

**Issue by mtjn** [microsoft/TypeScript#37382]
**TypeScript Version:**  Nightly

**Search Terms:** "argument not assignable" "optional property" 2345

**Expected behavior:** Typescript should realise that the property has been checked for truthiness and cannot be undefined, no error should be raised

**Actual behavior:** Typescript claims that the property could still be undefined and raises this error:

```
Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.(2345)
```

**Related Issues:**
https://github.com/microsoft/TypeScript/issues/37240
https://github.com/microsoft/TypeScript/issues/35964
https://github.com/microsoft/TypeScript/issues/36193

**Code**
```ts
interface IFoo {
  bar?: string;
}

type Obj = {[key: string]: IFoo}

function print(str: string) {
  return console.log(str);
}

export function check(obj: Obj, key: string) {
  if (obj[key]?.bar) {
    print(obj[key]?.bar);
  }
}

```

<details><summary><b>Output</b></summary>

```ts
function print(str) {
    return console.log(str);
}
export function check(obj, key) {
    var _a, _b;
    if ((_a = obj[key]) === null || _a === void 0 ? void 0 : _a.bar) {
        print((_b = obj[key]) === null || _b === void 0 ? void 0 : _b.bar);
    }
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](http://www.typescriptlang.org/play/?ts=3.9.0-dev.20200312&ssl=16&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgJIDED2nkG8BQyyARnFAPwBcyAzmFKAOYDc+AvvvmAJ4AOKAeWIArZAF48AbQDWEbtToMQjALrUM2DvhgBXEAjDBMIZLyVgAFIoX0mASjyFkUCGB1QTCYzUwAbCAB0vpiMVvR2rFoQAB68mFBgyLr6hsbICAAWEAjSFpgi1ELCADTIsvK0tsoOBETAMMh5IjJyKuQBpFA1TkRmoJb5wi3cbR1kEU4cHEA)

**Comment by RyanCavanaugh** [comment]
Duplicate #10530
